storyblok/storyblok-cli
# Add command for Typescript typedefs generation

## Pull request type

## How to test this PR
I have set up an npm `dev` script that can be useful to conveniently run the build step and the CLI in a "pseudo-dev mode".
- Log in
- Locate a file with a components JSON Schema (or generate one via the `pull-components` command)
- Run the `generate-typescript-typedefs` command providing the components JSON Schema via the `--sourceFilePaths` flag
- Inspect the generated Typescript definition file
- Try out different flags and check the output

### Command example
```
npm run dev generate-typescript-typedefs -- --sourceFilePaths ./components.12345.json
```

## What is the new behavior?
The main purpose of this PR is to add a command to the CLI that would provide our users with first-party support for generating Typescript type definitions based on Storyblok component schemas. As we're dealing with Typescript, it was a good opportunity to open up Typescript adoption in the CLI as well, as it made working with Typescript MUCH easier.

Here's a list of the provided changes:
- Create a `GenerateTypesFromJSONSchemas` class responsible for the whole rendering cycle of the Typescript type definitions file.
- Add support for Typescript files and build process with [`unbuild`](https://github.com/unjs/unbuild) - required updating all import types to be ESM
- Add `.vscode` settings to prevent automated formatting of existing `js` files while working on `ts` files
- Add a `generate-typescript-typedefs` CLI command and task file
- Add Typescript types for the whole task and its internals under the new directory `types`
- Add a section to the `README` file related to the `generate-typescript-typedefs` command

### TBD / Potential issues / Need hints
- As for now, the generated type definitions file is importing the `ISbStoryData` type from the same `storyblok` package, but there are some issues with exporting those types from this package, namely because some of them are also being imported from the `storyblok-js-client`. I'd assume that any user who is willing to use this new command is going to rely on the `storyblok-js-client`, at some point or another, so I guess the way to go is to somehow export the `types` alongside the `package.json` with `@storyblok/js` or `storyblok-js-client` listed as dependencies so that we can be sure we're going to import types that are going to be available from our distributed types. I still have to find out how to do that properly :D
- Maybe we need automated tests for this (read below)? There are a lot of options/combinations to handle, it's hard to test them all manually. Furthermore, updating CJS to ESM imports caused unit tests to broke, so I need to figure out how to fix them/refactor them.

### Potential improvements
As I had the chance to dig further into the CLI repo, I have outlined some potential improvements, for both the `generate-typescript-typedefs` command and the whole repo:
- [TS] Create a way to avoid clashes between `StoryblokProvidedPropertyTypes` and bloks (components) that can be created in Storyblok with the same name. I.e. a user can create a blok named `table`, and the types will end up having two `TableStoryblok` types, one for the user's component and one for the Storyblok-provided `Table` type.
- [TS] Give the possibility to the user to code the custom type parser in Typescript
- [TS] Add automated tests (need setup for TS tests) - Do you think this is required? It could also be a good opportunity to set them up with TS and to provide some mocks that could also be useful as internal examples, such as a `customFieldTypesParser` or a `JSONSchemaToTSOptions`.
- [CLI] Code formatting is missing
- [CLI] Unified error handling (maybe a dedicated class?)
- [CLI] Typescript support on all the codebase
- [CLI] There are undocumented commands, like `scaffold`, that are not documented in the README file. Maybe we could benefit from some cleanup. That command in particular I think is related to the old Rendering Service.

## Other information
A new `build` step should be added to trigger the `build` script.

This work has been heavily inspired by dohomi's [Storyblok generate ts](https://github.com/dohomi/storyblok-generate-ts). Thank you for your hard work and support! 🙏
